### PR TITLE
fix: esm.sh package lookups

### PR DESCRIPTION
**Pull request by guybedford**
This should fix unnecessary `package.json` requests for esm.sh by ensuring we are properly detecting package URLs for different CDN URL variations.

### CODE REVIEW

File: src/providers/esmsh.ts [hunk 1]
```diff
@@ -4,16 +4,16 @@ import { Resolver } from "../trace/resolver.js";
...
 export async function pkgToUrl(pkg: ExactPackage): Promise<`${string}/`> {
-  return `${cdnUrl}${pkg.name}@${pkg.version}/`;
+  // The wildcard '*' at the end tells the esm.sh CDN to externalise all
+  // dependencies instead of bundling them into the returned module file.
+  //   see https://esm.sh/#docs
+  return `${cdnUrl}*${pkg.name}@${pkg.version}/`;
 }
 
-const exactPkgRegEx = /^((?:@[^/\\%@]+\/)?[^./\\%@][^/\\%@]*)@([^\/]+)(\/.*)?$/;
+const exactPkgRegEx = /^(v\d+\/)?\*?((?:@[^/\\%@]+\/)?[^./\\%@][^/\\%@]*)@([^\/]+)(\/.*)?$/;
```
Comment (ije) [inline src/providers/esmsh.ts:16]: just tested the updated regexp, seems it returns incorrect name/version;

Screenshot 2023-05-03 at 19 56 11

fix patch:

```js
const [, , name, version] = url.slice(cdnUrl.length).match(exactPkgRegEx) || [];
```

File: src/providers/esmsh.ts [hunk 2]
```diff
@@ -4,16 +4,16 @@ import { Resolver } from "../trace/resolver.js";
...
 export async function pkgToUrl(pkg: ExactPackage): Promise<`${string}/`> {
-  return `${cdnUrl}${pkg.name}@${pkg.version}/`;
+  // The wildcard '*' at the end tells the esm.sh CDN to externalise all
+  // dependencies instead of bundling them into the returned module file.
+  //   see https://esm.sh/#docs
+  return `${cdnUrl}*${pkg.name}@${pkg.version}/`;
 }
 
-const exactPkgRegEx = /^((?:@[^/\\%@]+\/)?[^./\\%@][^/\\%@]*)@([^\/]+)(\/.*)?$/;
+const exactPkgRegEx = /^(?:v\d+\/)?\*?((?:@[^/\\%@]+\/)?[^./\\%@][^/\\%@]*)@([^\/]+)(\/.*)?$/;
```
Comment (ije) [inline src/providers/esmsh.ts:16]: nice `?:`

**Review by ije — APPROVED** [review]
LGTM, although don't know why the testing failed 🤔

**Comment by ije** [comment]
thanks

**Comment by guybedford** [comment]
Glad we could get it sorted out.

**Pull request merged**